File: src/main/java/io/github/tanguygab/tabadditions/shared/features/advancedconditions/SimpleCondition.java
package io.github.tanguygab.tabadditions.shared.features.advancedconditions;

import io.github.tanguygab.tabadditions.shared.TABAdditions;
import me.neznamy.tab.shared.TAB;
import me.neznamy.tab.shared.platform.TabPlayer;

import java.util.List;

/**
 * An abstract class representing a simple condition
 */
public abstract class SimpleCondition {

    /** Text on the left side of condition */
    protected String leftSide;

    /** Placeholders used on the left side */
    private final String[] leftSidePlaceholders;

    /** Text on the right side of condition */
    protected String rightSide;

    /** Placeholders used on the right side */
    private final String[] rightSidePlaceholders;

    public SimpleCondition(String[] arr) {
        leftSide = arr.length < 1 ? "" : arr[0];
        leftSidePlaceholders = TAB.getInstance().getPlaceholderManager().detectPlaceholders(leftSide).toArray(new String[0]);
        rightSide = arr.length < 2 ? "" : arr[1];
        rightSidePlaceholders = TAB.getInstance().getPlaceholderManager().detectPlaceholders(rightSide).toArray(new String[0]);
    }

    /**
     * Replaces placeholders on the left side and return result
     *
     * @return  replaced left side
     */
    public String parseLeftSide(TabPlayer viewer, TabPlayer target) {
        return parseSide(viewer, target, leftSide, leftSidePlaceholders);
    }

    /**
     * Replaces placeholders on the right side and return result
     * @return   replaced right side
     */
    public String parseRightSide(TabPlayer viewer, TabPlayer target) {
        return parseSide(viewer, target, rightSide, rightSidePlaceholders);
    }

    /**
     * Replaces placeholders in provided value
     *
     * @param   value
     *          string to replace placeholders in
     * @param   placeholders
     *          used placeholders
     * @return  replaced string
     */
    public String parseSide(TabPlayer viewer, TabPlayer target, String value, String[] placeholders) {
        return TABAdditions.getInstance().parsePlaceholders(value,viewer,target, List.of(placeholders));
    }

    /**
     * Returns {@code true} if condition is met for player, {@code false} if not
     *
     * @return  {@code true} if met, {@code false} if not
     */
    public abstract boolean isMet(TabPlayer viewer, TabPlayer target);
}

File: src/main/kotlin/io/github/tanguygab/tabadditions/shared/features/advancedconditions/SimpleCondition.kt
package io.github.tanguygab.tabadditions.shared.features.advancedconditions

import io.github.tanguygab.tabadditions.shared.TABAdditions
import me.neznamy.tab.shared.features.PlaceholderManagerImpl
import me.neznamy.tab.shared.platform.TabPlayer

/**
 * An abstract class representing a simple condition
 */
abstract class SimpleCondition(arr: List<String>) {
    /** Text on the left side of condition  */
    protected var leftSide: String = if (arr.isEmpty()) "" else arr[0]

    /** Placeholders used on the left side  */
    private val leftSidePlaceholders = PlaceholderManagerImpl.detectPlaceholders(leftSide)

    /** Text on the right side of condition  */
    protected var rightSide: String = (if (arr.size < 2) "" else arr[1])

    /** Placeholders used on the right side  */
    private val rightSidePlaceholders = PlaceholderManagerImpl.detectPlaceholders(rightSide)

    /**
     * Replaces placeholders on the left side and return result
     *
     * @return  replaced left side
     */
    fun parseLeftSide(viewer: TabPlayer, target: TabPlayer) = parseSide(viewer, target, leftSide, leftSidePlaceholders)

    /**
     * Replaces placeholders on the right side and return result
     * @return   replaced right side
     */
    fun parseRightSide(viewer: TabPlayer, target: TabPlayer) = parseSide(viewer, target, rightSide, rightSidePlaceholders)

    /**
     * Replaces placeholders in provided value
     *
     * @param   value
     * string to replace placeholders in
     * @param   placeholders
     * used placeholders
     * @return  replaced string
     */
    fun parseSide(viewer: TabPlayer, target: TabPlayer, value: String, placeholders: List<String>)
    = TABAdditions.INSTANCE.parsePlaceholders(value, target, viewer, placeholders)

    /**
     * Returns `true` if condition is met for player, `false` if not
     *
     * @return  `true` if met, `false` if not
     */
    abstract fun isMet(viewer: TabPlayer, target: TabPlayer): Boolean
}